### PR TITLE
Add timeout to in-flight telemetry reports

### PR DESCRIPTION
**Pull request by hawkw**
This PR adds a configurable timeout duration after which in-flight telemetry reports are dropped, cancelling the corresponding RPC request to the control plane. 

Right now, the default timeout is 10 seconds, but I'm not sure if this is a reasonable default value. I'd also like to add a unit test ensuring that this works.

### CODE REVIEW

**Comment by olix0r** [comment]
I wonder if we can do this in a more generic way so that timeouts can be added to arbitrary requests...

@carllerche @seanmonstar do you have any thoughts on what tower-timeout might look like? ;)

**Comment by seanmonstar** [comment]
A basic version could be put next to the `Backoff` impl that's currently in tree, before we figure out exactly what tower-timeout would look like. I think the changes to timers would affect our design, so it's likely faster to just do hardcoded to the current `tokio_core::reactor::Timeout` for now...

However, I do agree that a generic solution would be best: having a `Timeout` service wrap another.

**Comment by hawkw** [comment]
> I wonder if we can do this in a more generic way so that timeouts can be added to arbitrary requests...
I wondered about that as well; in this case it seemed like since we already had the timestamp it was simpler to just add the if here. Can definitely write a generic timeout implementation, though.

**Comment by hawkw** [comment]
Okay, @seanmonstar and @olix0r, what do you think of 32e33c43a116ba97685afc8b996439d142845cac?

File: proxy/src/timeout.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        } else if let Async::Ready(_) = self.timeout.poll().expect("timer failed") {
+            Err(TimeoutError::Timeout(self.duration))
+        } else {
+            Ok(Async::NotReady)
+        } 
+    }
+}
+
+// We have to provide a custom implementation of Debug, because
+// tokio_core::reactor::Timeout is not Debug.
```
Comment (briansmith) [inline proxy/src/timeout.rs]: We really need to stop doing this. What's stopping the new new version of `tokio_core` that includes this fix?

File: proxy/src/config.rs [hunk 2]
```diff
@@ -100,6 +103,7 @@ pub enum UrlError {
 // Environment variables to look at when loading the configuration
 const ENV_EVENT_BUFFER_CAPACITY: &str = "CONDUIT_PROXY_EVENT_BUFFER_CAPACITY";
 const ENV_METRICS_FLUSH_INTERVAL_SECS: &str = "CONDUIT_PROXY_METRICS_FLUSH_INTERVAL_SECS";
+const ENV_REPORT_TIMEOUT_SECS: &str = "CONDUIT_PROXY_REPORT_TIMEOUT_SECS";
```
Comment (briansmith) [inline proxy/src/config.rs]: If we have timeouts expressed as seconds then we can't write tests for this that time out in less than a second without the tests skipping over the `Config` structure's parsing of the settings strings, which reduces the test coverage of the settings strings parsing.

This is also inconsistent with other timeouts like `ENV_PRIVATE_CONNECT_TIMEOUT` and `ENV_PUBLIC_CONNECT_TIMEOUT` which are set in milliseconds already.

**Pull request merged**